staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go

```diff
@@ -0,0 +1,274 @@
+	}
+	testCustomResourceConversionWebhook(t, crd, versionedClients)
+	testCRListConversion(t, crd, apiExtensionClient, versionedClients)
+}
+
+func verifyV1Object(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, obj *unstructured.Unstructured) {
+	assert.Equal(t, crd.Spec.Group+"/v1", obj.GetAPIVersion())
+	hostPort, exists := obj.Object["hostPort"]
+	assert.True(t, exists)
+	assert.Equal(t, "localhost:8080", hostPort)
+	_, hostExists := obj.Object["host"]
+	assert.False(t, hostExists)
+	_, portExists := obj.Object["port"]
+	assert.False(t, portExists)
+}
+
+func verifyV2Object(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, obj *unstructured.Unstructured) {
+	assert.Equal(t, crd.Spec.Group+"/v2", obj.GetAPIVersion())
+	_, hostPortExists := obj.Object["hostPort"]
+	assert.False(t, hostPortExists)
+	host, hostExists := obj.Object["host"]
+	assert.True(t, hostExists)
+	assert.Equal(t, "localhost", host)
+	port, portExists := obj.Object["port"]
+	assert.True(t, portExists)
+	assert.Equal(t, "8080", port)
+}
+
+func testCustomResourceConversionWebhook(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, customResourceClients map[string]dynamic.ResourceInterface) {
+	name := "cr-instance-1"
+	t.Log("Creating a v1 custom resource")
+	crInstance := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       crd.Spec.Names.Kind,
+			"apiVersion": crd.Spec.Group + "/v1",
+			"metadata": map[string]interface{}{
+				"name": name,
+			},
+			"hostPort": "localhost:8080",
+		},
+	}
+	_, err := customResourceClients["v1"].Create(crInstance, metav1.CreateOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	t.Log("v2 custom resource should be converted")
+	v2cr, err := customResourceClients["v2"].Get(name, metav1.GetOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	if !assert.NotNil(t, v2cr) {
+		t.FailNow()
+	}
+	verifyV2Object(t, crd, v2cr)
+}
+
+func testCRListConversion(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, apiextensionsClient apiextensionsclientset.Interface, customResourceClients map[string]dynamic.ResourceInterface) {
+	name1 := "cr-instance-1"
+	name2 := "cr-instance-2"
+	t.Log("Creating a v1 custom resource")
+
+	// cr-instance-2 is stored as v1. Now change storage version to v2.
+	var updateCRD *apiextensionsv1beta1.CustomResourceDefinition
+	var err error
+	for i := 0; i < 5; i++ {
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to get CustomResourceDefinition %q: %v", crd.Name, err)
+		}
+
+		// swap storage version
+		updateCRD.Spec.Versions[0].Storage = false
+		updateCRD.Spec.Versions[1].Storage = true
+
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Update(updateCRD)
```

Comment:
do we need to wait for the handler to update?

Reply:
yes. How can we without trying again and again while checking etcd or by checking via the webhook?

Reply:
Added the TODO at https://github.com/kubernetes/kubernetes/pull/77868/files#diff-7dafb2a7b5eb7c3edad4fdc93db9a243R179. https://github.com/kubernetes/kubernetes/pull/77756/files#diff-7dafb2a7b5eb7c3edad4fdc93db9a243R350 brings a variant with etcd access code. We should use that.

Reply:
fixed by rebasing on @jpbetz's test framework.